src/quote/provider/icusd.ts

```diff
@@ -0,0 +1,136 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { getTokenByChainAndSymbol } from '@indexcoop/tokenlists'
+
+import {
+  FlashMintWrappedBuildRequest,
+  WrappedTransactionBuilder,
+} from 'flashmint'
+
+import { WrappedQuoteProvider } from '../flashmint/wrapped'
+import { QuoteProvider, QuoteToken } from '../interfaces'
+import {
+  FlashMintContractType,
+  FlashMintQuote,
+  FlashMintQuoteRequest,
+} from '../provider'
+import { SwapQuoteProvider } from '../swap'
+import { buildQuoteResponse } from './utils'
+
+export interface IcUsdQuoteRequest extends FlashMintQuoteRequest {
+  chainId: number
+  isMinting: boolean
+  inputToken: QuoteToken
+  outputToken: QuoteToken
+  indexTokenAmount: BigNumber
+  // FlashMintNav always needs the input amount
+  inputTokenAmount: BigNumber
+  slippage: number
+}
+
+export class IcUsdQuoteRouter
+  implements QuoteProvider<IcUsdQuoteRequest, FlashMintQuote>
+{
+  constructor(
+    private readonly rpcUrl: string,
+    private readonly swapQuoteProvider: SwapQuoteProvider
+  ) {}
+
+  async getQuote(request: IcUsdQuoteRequest): Promise<FlashMintQuote | null> {
+    const { chainId } = request
+    const usdc = getTokenByChainAndSymbol(chainId, 'USDC')
+    const icUsd = getTokenByChainAndSymbol(chainId, 'icUSD')
+    if (!usdc || !icUsd) {
+      throw new Error(`icUSD not supported on chainId: ${chainId}`)
+    }
+    // For now we only use FlashMintWrapped
+    // if (request.isMinting) {
+    //   return await this.getFlashMintNavQuote(request)
+    // } else {
+    //   const usdcBalance = await getBalanceOf(
+    //     usdc.address,
+    //     icUsd.address as Address,
+    //     chainId
+    //   )
+    //   const usdcInputAmount = await getExpectedReserveRedeemQuantity(
+    //     icUsd.address as Address,
+    //     usdc.address,
+    //     indexTokenAmount.toBigInt()
+    //   )
+    //   console.log(usdcBalance.toString(), 'USDC')
+    //   // 80% of the USDC balance of icUSD
+    //   const threshold = (usdcBalance * 80n) / 100n
+    //   const useFlashMintNav = usdcInputAmount < threshold
+    //   if (useFlashMintNav) return await this.getFlashMintNavQuote(request)
+    return await this.getFlashMintWrappedQuote(request)
```

Comment:
For now FlashMintWrapped only. Leaving routing logic to be able to go back to this if needed soon.